Add-ons: Disambiguations for dither layers and registries

Noordfrees opened the pull request:
- Fix #4897 
- Added some extra info to description registries which will be useful for future enhancements as well
- Print warnings about potentially unsafe batch registries, and fixes batch registries for the official resources. Some immovables (tree, rocks, fields etc) still use it, this is something for a separate branch.

Noordfrees committed:
Refactor dither layer handling

Noordfrees committed:
Add warning about batch-registering units

Noordfrees committed:
Demultiply resource registries

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
5 files were automatically formatted.

Noordfrees committed:
fix doc path

Noordfrees committed:
Fix testsuite

Noordfrees committed:
fix testsuite

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
'src/logic/map_objects/description_manager.cc' was automatically formatted.

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
'src/scripting/lua_map.cc' was automatically formatted.

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

frankystone committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

Noordfrees committed:
Merge branch 'master' into terrain-addons-enhancements

matthiakl reviewed (commented):
Seems fine. Some comments inline.
I did not test it yet (not sure how?).

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into terrain-addons-enhancements

Noordfrees committed:
Address review

Noordfrees commented:
The terrain changes can be tested by creating two world add-ons with new terrains that use the same dither layer, and checking in the editor that there are no sharp edges between any two terrains.

For the registries, this can be checked by running Widelands with `--verbose`, which should print some warnings in the output – some for official units, and even more if add-ons with batch registries are enabled.

klaus-halfmann commented:
I tryed to checkout your branch via my common repo;
```
git remote -v
nf	git@github.com:Noordfrees/widelands.git (fetch)
nf	git@github.com:Noordfrees/widelands.git (push)
origin	git@github.com:widelands/widelands.git (fetch)
origin	git@github.com:widelands/widelands.git (push)
upstream	git@github.com:widelands/widelands.git (fetch)
upstream	git@github.com:widelands/widelands.git (push)
```
But `git checkout nf/terrain-addons-enhancements` fails me 

Same for clone
```
$ git clone repo -o nf -b nf/terrain-addons-enhancements terrain-addons-enhancements
Cloning into 'terrain-addons-enhancements'...
fatal: Remote branch nf/terrain-addons-enhancements not found in upstream nf
```
This fails, too:
`git clone repo -o nf -b remotes/nf/terrain-addons-enhancements  terrain-addons-enhancements `

However the Branch should be here:
```
repo> git branch --list --all | fgrep terrain                         
warning: ignoring broken ref refs/remotes/upstream/HEAD
  remotes/nf/terrain-addons-enhancements
```

Going back to the "normal" clone for now

Noordfrees commented:
What does `git remote show nf` on your common repo show? It should say something like this:
```
benedikt@benediktPC:~/wl/widelands$ git fetch --prune nf 
benedikt@benediktPC:~/wl/widelands$ git remote show nf 
* Remote-Repository nf
  URL zum Abholen: git@github.com:Noordfrees/widelands.git
  URL zum Versenden: git@github.com:Noordfrees/widelands.git
  Hauptbranch: master
  Remote-Branches:
    4184-move-workareas_feature-rejected gefolgt
    boost-format                         gefolgt
    diplomacy                            gefolgt
    fri-build-anims                      gefolgt
    fri03_old                            gefolgt
    fri04                                gefolgt
    infinite_production                  gefolgt
    lua-style-checker                    gefolgt
    master                               gefolgt
    misc-addon-enhancements              gefolgt
    more-clangtidy-checks                gefolgt
    multithreading_old                   gefolgt
    naval_warfare_old                    gefolgt
    navalwarfare                         gefolgt
    rip-out-boost-1                      gefolgt
    scenario-editor_old                  gefolgt
    terrain-addons-enhancements          gefolgt
  Lokaler Branch konfiguriert für 'git pull':
    ...
  Lokale Referenzen konfiguriert für 'git push':
    ...
```
You can also try `git checkout terrain-addons-enhancements` – sometimes this works and sometimes it doesn't ;)

klaus-halfmann reviewed (changes requested):
I have no real idea how to test hits:
* Open the Editor and examin all possible terrain permutations?
* Open as many (terrain) addons as possible and then ?

What happens if we reach the unit_16 maximum, will we just get an overflow?
Please elaborate more on:
* what is your solution to the dither_layer problem
* why did you move the resouces into a different path e.g. pics->coal
I compiled this now and will try something along my ideas.
Could we have a jitsi call tommorow at perhpas 18:00?

Some comments inline, maybe you can fix them meanwhile

klaus-halfmann commented:
When I load Add-Ons, I get warnings from the already installed ones, this is confusing.
I even get Warnings when I apply some config change?
When I load a savgame I get:
`00:00:15.968 real] ERROR:   FATAL EXCEPTION: messages for player 1: "Water": [/Users/klaus/develop/widelands/terrain-addons-enhancements/src/graphic/image_io.h:34] Image not found: world/resources/pics/water4.png`

Does _this_ change break compatibily or is this much older then ["Along the river"](https://github.com/widelands/widelands/pull/5062) ?

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into terrain-addons-enhancements

Noordfrees committed:
address review

Noordfrees committed:
Revert renaming

Noordfrees committed:
'src/logic/map_objects/descriptions.cc' was automatically formatted.

Noordfrees commented:
I renamed the resource images because @matthiakl suggested it, but this broke savegame compatibility, so I reverted it now. The savegame should load again.

---

### Dither layers.
Every terrain has a dither layer. This is a 32-bit signed integer that is required to be unique across all terrains. The dither layer determines which terrain is dithered on top of the other one at edges where two terrains meet. In master, if you enable two world add-ons that both define a terrain with the same dither layer, Widelands will fail to start the editor. In this branch, dither layer are multiplied with a constant (`kMaxDitherLayerDisambiguator`) and an offset added to ensure that no such clashes are possible, ever. A terrain's dither layer is now calculated as 
```c++
dither_layer = dl * kMaxDitherLayerDisambiguator + off
```
where `dl` is the value from the terrain's `init.lua` and `kMaxDitherLayerDisambiguator` an arbitrary constant. `off` is 0 for official terrains, 1 for the first enabled add-on, 2 for the second enabled add-on, and so on. As long as fewer than `kMaxDitherLayerDisambiguator` add-ons are enabled and no add-on defines two terrains with the exact same layer, this guarantees that there will be no clashes. Problems can only occur if:
- The player enables 100 (`kMaxDitherLayerDisambiguator`) or more add-ons (we just have to assume nobody does that; the limit is arbitrary and could be increased)
- An add-on author creates an add-on with two terrains that use the exact same layer. This is forbidden, and modders are told not to do this by the documentation.
- An add-on author specifies a dither layer `dl` so large that `dl * kMaxDitherLayerDisambiguator` overflows `int32_t`. Let's assume modders are responsible enough not to try such crazy values.

---

To test this change, it should actually be enough to launch the editor with some world add-ons enabled. Especially add-ons in which you edited the dither layers so two different add-ons use the same layer. In master, this configuration will fail; in this branch, it should load fine.

---

> When I load Add-Ons, I get warnings from the already installed ones, this is confusing.
I even get Warnings when I apply some config change?

When you change your add-ons configuration, Widelands performs a quick check whether your configuration is valid, and displays an error message if this is not the case. So any bad practices are logged every time you do this, and also every time you start a game or the editor. Warnings are currently shown for several existing add-ons, and also a few official units which should be cleaned up in a follow-up branch.

>  Could we have a jitsi call tommorow at perhpas 18:00?

:+1:

klaus-halfmann commented:
Thanks for that explanation. Can I find that as comments in the code or some documentation, too?

Can I just duplicate and rename some existing AddOn? This should provoke the situation.

Savegemag works again , found this Crash here, too: #5111

klaus-halfmann commented:
I duplicated the `foreign_planet.wasd` but this branch and master has no Issues?

klaus-halfmann commented:
Lets meet here: https://meet.ffmuc.net/WidelandsChat

Noordfrees commented:
> I duplicated the `foreign_planet.wasd` but this branch and master has no Issues?

You need to rename the terrains in the copy, otherwise the terrains of one add-on are silently ignored. (That's a feature, not a bug, because the add-on itself gets to define how to behave in such a case, and my foreign_planet is very well-behaved :wink:)

klaus-halfmann commented:
I modified terrains/init.lua   but master still has no Issue, lets check this in tomorrows call.

matthiakl commented:
I got it to work with foreign_planet and implassable_water without problems.

However, when I try to hack something together by copying one addon and just change some terrain names, so they get loaded, it looks like the dither layer check is performed before evaluating the __skip_if_exists attribute:
```
[00:00:12.670 real] ERROR: ##############################
[00:00:12.670 real] ERROR:   FATAL EXCEPTION in editor: [/mnt/storage/Programme/widelands-dev/src/scripting/lua_errors.cc:22] Terrain foreignplanet_rocks1 has the same dither layer 35301 as foreignplanet_rocks1
[00:00:12.670 real] ERROR: stack traceback:
[00:00:12.670 real] ERROR: 	[string "addons/foreign_planet.wad/terrains/in..."]:8: in main chunk
[00:00:12.670 real] ERROR: ##############################
[00:00:31.155 real] WARNING: The registry file addons/foreign_planet.wad/terrains/register.lua
[00:00:31.155 real] WARNING:     defines multiple (14) units: foreignplanet_beach foreignplanet_dust1 foreignplanet_dust2 foreignplanet_mountain1 foreignplanet_mountain2 foreignplanet_mountain3 foreignplanet_rocks1 foreignplanet_rocks2 foreignplanet_rocks3 foreignplanet_rocks4 foreignplanet_rocks5 foreignplanet_rocks6 foreignplanet_water1 foreignplanet_water2
[00:00:31.155 real] WARNING:     Defining multiple units in one file is error-prone. Every unit should be defined in a registry directory of its own.
[00:00:31.157 real] INFO: Skipped registering 'foreignplanet_beach'
[00:00:31.157 real] INFO: Skipped registering 'foreignplanet_dust1'
[00:00:31.157 real] INFO: Skipped registering 'foreignplanet_dust2'
[00:00:31.157 real] INFO: Skipped registering 'foreignplanet_mountain1'
[00:00:31.157 real] INFO: Skipped registering 'foreignplanet_mountain2'
[00:00:31.157 real] INFO: Skipped registering 'foreignplanet_mountain3'
[00:00:31.157 real] INFO: Skipped registering 'foreignplanet_rocks1'
[00:00:31.157 real] INFO: Skipped registering 'foreignplanet_rocks2'
[00:00:31.157 real] INFO: Skipped registering 'foreignplanet_rocks3'
[00:00:31.157 real] INFO: Skipped registering 'foreignplanet_rocks4'
[00:00:31.157 real] INFO: Skipped registering 'foreignplanet_rocks5'
[00:00:31.157 real] INFO: Skipped registering 'foreignplanet_rocks6'
```

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into terrain-addons-enhancements

Noordfrees commented:
I cannot reproduce this… **however** I just noticed that my foreign_planet does not follow best practices (somehow got overlooked or whatever – will prepare an update this evening) so you need to rename **all** terrains in this add-on, in all of the three files `init.lua` `register.lua` `editor.lua` for this duplication to create the situation that fails in master and works here.

Noordfrees committed:
Reduce memory reallocations

matthiakl reviewed (approved):
Ok, so I should stop messing around with add-ons which I don't know ;)
From my side this is fine now. 

@klaus-halfmann, I set a terrain in impassable_water to the same layer as one in foreign_planet. As expected it fails in master, but works with this branch. Depending on the load order one or the other terrain is shown above.

klaus-halfmann reviewed (approved):
Did a review with @Noordfrees

klaus-halfmann commented:
Tetsing this  
[foreign_planet2.wad.zip](https://github.com/widelands/widelands/files/7532468/foreign_planet2.wad.zip)
results in  
`assertion failed: (registered_scenario_descriptions_.count(description_name) == 1 || registered_descriptions_.count(description_name) == 1), function mark_loading_in_progress, file description_manager.cc, line 267.`

Noordfrees commented:
Thank you both for the reviews :) 

For the record the assert failure turned out to be due to a very corrupted add-on and not directly related to this branch.
